Engine/KDtree.h

```diff
@@ -0,0 +1,503 @@
+		std::vector<Triangle*> tris;
+		Triangles(){
+			
+		}
+		//move	
+		Triangles(Triangles&& other) {
+			//tris.swap(other.tris);
+			tris = std::move(other.tris);
+		}		
+		//implement initilize list
+		Triangles& operator=(Triangles&& other) {
+			//tris.swap(other.tris);
+			tris = std::move(other.tris);
+			return *this;
+		}
+
+		//copy
+		Triangles(const Triangles& other) {
+			//tris.swap(other.tris);
+			tris = other.tris;
+		}		
+
+		Triangles& operator=(const Triangles& other) {
+			//tris.swap(other.tris);
+			tris = other.tris;
+			return *this;
+		}
+		
+};
+class Voxel;
+class KDNode: public Primitive{
+	public:
+		KDNode *left;
+		KDNode *right;
+		Triangles Tprim;
+		Plane plane;
+   		KDNode(const PrimitiveMetadata* metadata) : Primitive(metadata)
+    	{ 
+        	left = NULL; 
+        	right = NULL; 
+    	} 
+		bool isLeaf(){
+			return (left==NULL && right==NULL);
+		}
+		
+		bool isIntersecting(const Ray& ray, HitProbe& probe) const override;
+		void calcIntersectionDetail(const Ray& ray, HitProbe& probe, HitDetail* out_detail)  const override;
+		bool isIntersectingVolumeConservative(const AABB3D& volume) const override;
+		void calcAABB(AABB3D* out_aabb) const override;
+		KDNode* recBuild(Triangles& T, Voxel& V, int depth);
+		KDNode* build_KD_tree(Triangles& T);
+};
+class Voxel{
+	public:
+		AABB box;
+		Voxel(){
+
+		}
+		int LongestAxis(){
+			if(box.is_set()){		
+				Vec3 temp = box.getVertexMax() - box.getVertexMin();
+				if(temp.x > temp.y && temp.x > temp.z){
+					return AXIS_X;
+				}
+				else if(temp.y > temp.x && temp.y > temp.z){
+					return AXIS_Y;
+				}
+				else{
+					return AXIS_Z;
+				}
+			}
+
+			fprintf(stderr,"set AABB box of Voxel first, use copyVertexMin/Max or set VertexMin/Max");
+			exit(1);
+		}
+		bool intersect(const Ray& ray, Voxel World_Voxel,float *outTMin, float *outTMax){
+			//base by pbrt			
+			float t0 = 0;
+			float t1 = ray.getMaxT();
+			Vector3R temp = ray.getDirection();
+			float dir[3];
+			dir[0] = temp.x;
+			dir[1] = temp.y;
+			dir[2] = temp.z;
+			
+			temp = ray.getOrigin();
+			float origin[3];
+			origin[0] = temp.x;
+			origin[1] = temp.y;
+			origin[2] = temp.z;
+
+			float boxMax[3];
+			float boxMin[3];
+			boxMin[0] = World_Voxel.box.getVertexMin().x;
+			boxMin[1] = World_Voxel.box.getVertexMin().y;
+			boxMin[2] = World_Voxel.box.getVertexMin().z;
+
+			boxMax[0] = World_Voxel.box.getVertexMax().x;
+			boxMax[1] = World_Voxel.box.getVertexMax().y;
+			boxMax[2] = World_Voxel.box.getVertexMax().z;						
+
+			for(int i = 0; i < 3; i ++){
+				float inv = 1/dir[i];
+				float tNear = inv * (boxMin[i] - origin[i]);
+				float tFar = inv * (boxMax[i] - origin[i]);
+				if(tNear>tFar){
+					std::swap(tNear,tFar);
+				}
+			
+				t0 = tNear > t0 ? tNear : t0;
+   				t1 = tFar  < t1 ? tFar  : t1;
+   				if (t0 > t1) return false;
+			}
+			 if(outTMin) *outTMin = t0;
+			 if(outTMax) *outTMax = t1;
+			 return true;
+		}
+};
+
+
+typedef struct s{
```

Comment:
An in-class declaration or a dedicated class is encouraged. Using declaration is also preferred, too.
See: https://en.cppreference.com/w/cpp/language/using_declaration